lib/internal/streams/pipeline.js

```diff
@@ -63,8 +69,84 @@ function popCallback(streams) {
   return streams.pop();
 }
 
+function isPromise(obj) {
+  return !!(obj && typeof obj.then === 'function');
+}
+
+function isReadable(obj) {
+  return !!(obj && typeof obj.pipe === 'function');
+}
+
+function isWritable(obj) {
+  return !!(obj && typeof obj.write === 'function');
+}
+
+function isStream(obj) {
+  return isReadable(obj) || isWritable(obj);
+}
+
+function isIterable(obj, isAsync) {
+  if (!obj) return false;
+  if (isAsync === true) return !!obj[SymbolAsyncIterator];
+  if (isAsync === false) return !!obj[SymbolIterator];
+  return !!(obj[SymbolAsyncIterator] || obj[SymbolIterator]);
+}
+
+function makeAsyncIterable(val, name) {
+  if (isIterable(val)) {
+    return val;
+  } else if (isReadable(val)) {
+    return _fromReadable(val);
+  } else {
+    throw new ERR_INVALID_RETURN_VALUE(
+      'AsyncIterable, Readable', name, val);
+  }
+}
+
+async function* _fromPromise(val) {
+  yield await val;
+}
+
+async function* _fromReadable(val) {
+  if (isReadable(val)) {
+    // Legacy streams are not Iterable.
+
+    if (!Readable) {
+      Readable = require('_stream_readable');
+    }
+
+    try {
+      const it = Readable.prototype[SymbolAsyncIterator].call(val);
```

Comment:
I guess this is more of a point for the readable async iterator implementation, but do you know if there’s a test somewhere testing how generic this method is? It seems like it only supports streams that have a `_readableState` property, i.e. no streams 1 streams?

Reply:
It's not very generic. It requires at least `'readable'` + `read`. Not sure which version of streams that is... we might consider to improve on this.

Is this something you would like to address in this PR? Or maybe add a doc note somehow?

Reply:
Err, yeah, that would not be something for this PR, rather a TODO comment or something like that…

Reply:
Fixed :D. https://github.com/nodejs/node/pull/31223/commits/ec7f64d4349c78c6c39d3726df18fa0f702484b7